zerver/views/auth.py

```diff
@@ -92,6 +95,14 @@ def maybe_send_to_registration(request: HttpRequest, email: str, full_name: str=
     depending on is_signup, whether the email address can join the
     organization (checked in HomepageForm), and similar details.
     """
+    assert not (mobile_flow_otp and desktop_flow_otp)
+    if mobile_flow_otp:
+        set_expirable_session_var(request.session, 'registration_mobile_flow_otp', mobile_flow_otp,
+                                  expiry_seconds=3600)
+    elif desktop_flow_otp:
+        set_expirable_session_var(request.session, 'registration_desktop_flow_otp', desktop_flow_otp,
+                                  expiry_seconds=3600)
```

Comment:
Would it be cleaner to attach these as properties on the `PreregistrationUser` object associated with the `create_confirmation_link` call here?

Reply:
It crossed my mind, but I saw two issues:
1. If someone ends up finishing the registration at a later time, they may be doing it on another device (not on their mobile anymore for example). Though that could be fixed by having these properties with an expiry time too. But still, session seems nice by tying this to the device.

2. We'll lose the information if the user tries a wrong email address and needs to try again:
```
    # This email address it not allowed to join this organization, so
    # just send the user back to the registration page.
    url = reverse('register')
    context = login_context(request)
    extra_context = {'form': form, 'current_url': lambda: url,
                     'from_multiuse_invite': from_multiuse_invite,
                     'multiuse_object_key': multiuse_object_key}  # type: Mapping[str, Any]
    context.update(extra_context)
    return render(request, 'zerver/accounts_home.html', context=context)
```
or we'll need to plumb these two params through here as well - session allows us to have a single way of transporting them.

Reply:
> If someone ends up finishing the registration at a later time, they may be doing it on another device (not on their mobile anymore for example).

I'm not sure that's a real issue; is there a way to transfer an in-progress registration between devices?  I think at least it could be solved.

> We'll lose the information if the user tries a wrong email address and needs to try again:

I guess the "wrong email address" detail is a good argument for why it should work this way; we don't know which PreregistrationUser to attach it to (unless it's responding to an invitation) in that case.  

The potential downside of storing something in the session is it could affect the next registration you do in your browser if you don't complete the one you began here, similar to the social auth bug we just fixed?

Reply:
> I guess the "wrong email address" detail is a good argument for why it should work this way; we don't know which PreregistrationUser to attach it to (unless it's responding to an invitation) in that case.

True, but from user experience - one will go to signup from mobile, make a mistake and try an email that's not allowed to join - they realize they have to try with their work email, so they do that - and end up logged in the mobile browser instead of the app and be annoyed.

> The potential downside of storing something in the session is it could affect the next registration you do in your browser if you don't complete the one you began here, similar to the social auth bug we just fixed?

Yes, it's an annoyance, the idea of the expiry of this session variable is to mitigate that, on the assumption that if the user's does the sign up on this device within X time (maybe 1hour that I set it to is too much), they're most likely to want to go to the app. And if the user makes the account, and then wants to register another one in the browser, that won't be "polluted", because we clear the ``otp`` from the session on succesful signup completion:
```
def login_and_go_to_home(request: HttpRequest, user_profile: UserProfile) -> HttpResponse:
    mobile_flow_otp = get_expirable_session_var(request.session, 'registration_mobile_flow_otp',
                                                delete=True)
    desktop_flow_otp = get_expirable_session_var(request.session, 'registration_desktop_flow_otp',
                                                 delete=True)
```

Those have been my reasoning for the session approach. Should we open a thread in #backend for this, or just go with either PreregUser/session approach?

Reply:
I think the session approach is fine; we should just add a nice block comment explaining why we use this different way of passing data around for this specific field.